### ISSUE TITLE
lll: Ignore long URLs, strings, and struct tags

### ISSUE THREAD

**Issue by mitar** [golangci/golangci-lint#3983]
### Your feature request related to a problem? Please describe.

There are upstream issues in lll about those with very good arguments:

* https://github.com/walle/lll/issues/12
* https://github.com/walle/lll/issues/11

### Describe the solution you'd like.

I think ignoring those should at least be possible through options, but I would even claim that those should be defaults. There is simply no easy way to wrap URLs and struct tags. And wrapping long strings just makes it less readable.

### Describe alternatives you've considered.

I currently have to manually add lint ignore pragmas in those cases.

### Additional context.

Upstream lll repository seems not maintained. So I am opening an issue here. There is a precedent for this, see upstream issue https://github.com/walle/lll/issues/15 and fix here https://github.com/golangci/golangci-lint/pull/3288.

**Comment by ldez** [comment]
lll is based on text content, there is no AST, so this linter doesn't understand the Go syntax.

**Comment by mitar** [comment]
I mean, if it simply does not count long strings of any type to the line length, I think this is then enough? So it just counts number of characters in a line which are outside of any strings. That handles struct tags as well.

**Comment by ldez** [comment]
a string is a language-related notion.

**Comment by ldez** [comment]
Theoretically and historically, lll is a linter that should limit line length to fit the length of a screen.

In theory, this linter should not care about language-related notions except if there is a compilation problem (like `import`) because the target is the length of the screen.

For me, the 2 issues (URL and string) are not related at all: one is about comment, and one is about string.

The following example (split of the URL) results in a valid link inside the render of godoc:
```go
// main runs the app.
// https://github.com/walle/lll/blob/4438bccd245f
// 7e87a5a69023625f01e7035c05c0/utils.go#L15
func main() {
```

I know it feels weird but godoc ignores `\n`.

In the modern world, the screen length is not really a problem: screens are no longer limited to 80 characters and IDEs have line wrapping.

My opinion and it's just my opinion, the rule behind lll is something from the past and not really relevant in the modern world. I prefer [semantic line breaks](https://sembr.org/) inside documentation and comment.

**Comment by ldez** [comment]
About the `string`, for me, the 2 points of view (not splitting or splitting) illustrate the problem behind `lll`: never splitting a long line is not a good rule, and always splitting a long line is not a good rule.
The good rule is to split or not depending on the context, a linter cannot help because a human must take a decision.

The current approach that limiting by default and can be ignored by `nolint` is the best approach because it relies on a human decision.

**Comment by ldez** [comment]
I created PR #3986 to ignore comments with only an URL.

But as explained in my previous comments, the other topics (`string` and `struct tag`) will not be handled.

**Comment by mitar** [comment]
Not sure why lll would not also ignore lines with backtick strings in them. You can have a regular expression to do that. And that would cover both struct tags and long strings. That addresses I think your comments/views as well: let's have backtick strings be allowed and not counted towards the length of the line while regular ones do count. Users can then decide which one to use as well as a way to communicate the context.

> About struct tag, this is a language-related notion, it's not manageable without the AST.
> The best solution is to use nolint directive in this context.

I disagree here a lot. We are talking about Go linter - lll. Not some general line length linter for general files. I think this is the main difference here. I would expect that all linters in golangci-lint know and understand Go and specially handle Go specifics (imports, struct tags, etc.). There is simply no case where you would not want to disable struct tags line length. So making this something that all users of golangci-lint have to set to ignore is simply ridiculous.

Now you are saying that this is only doable with AST. Then maybe the linter should also use AST to do its work. If it is too simple how it currently operates it is too simple. That is the beauty of Go, that AST is easy available and that you can do all kinds of linters. Otherwise I can just have vim rewrap text for me without understanding the language. But linter becomes useful because it can understand the language.

But in my view simply removing all backtick strings before counting the line length would fix the issue in my view. But if you think strings should not be touched, then we need AST to solve the struct tags.

**Comment by bombsimon** [comment]
I think struct tags is actually a valid thing to filter out because that's the only case that's not even supported by the language (there's no way to split over multiple lines) but I don't think it's that simple without the AST since you can put one or more comments anywhere in such line. This is valid Go code:

```go
type A struct {
    /* Name */ Name /* Type */ int /* Tag */ `tag:"value"`
}
```

And I don't like the idea of false positives or false negatives which would occur by ignore such (edge) cases.

Ignoring lines with a backtick (`` ` ``) is somewhat the same.

```go
var x = `very [...] long line` // Would be ignored

var y = `very
[...] long
line` // Would yield error on the second line if too long
```

I'm not sure if I think #3986 is a good idea, it looks like a very specific use case and might just be confusing to allow a single URL but not something like the comment suggests of `See: [url]`.

Not really related by just to confirm this is a personal and opinionated topic, I think `lll` is a reasonable rule to have even with modern monitors. I prefer wrapped lines because it's easier for me to read and it also makes me able to read code even if I use multiple splits in my editors (which I almost always do).

**Comment by ldez** [comment]
> I think struct tags are actually a valid thing to filter out

I didn't say the opposite but this issue is trying to mix 3 different topics.
I think that comment with only a URL is easy to handle and doesn't require complex regular expression.

I don't want to create a magic pattern (`See: `).

Your syntax is not complete:
```go
type A struct {
    /* Name */ Name /* Type */ int /* Tag */ `tag:"value"`
}
```

This case is also valid:
```go
type A struct {
    /* Type */ int /* Tag */ `tag:"value"`
}
```

And to detect that without a false positive, you have to know that you are in a struct.


> Not really related by just to confirm this is a personal and opinionated topic, I think lll is a reasonable rule to have even with modern monitors.

My opinion is that a rule simply based on length is not relevant, the split must be based on semantics and not length.

**Comment by mitar** [comment]
@ldez Thank you for being open in changing `lll` in ways you personally do not necessary agree with.

I am not familiar with AST internals and how hard it is to line length counting using AST, but I could imagine an ugly tool which would simply reformat a Go file and remove all struct tags. And then count line lengths in the resulting file. Any line which is too long in those reformatted file would be reported as too long in the original file. Reformatted file is just a temporary (or even in-memory) file.

**Comment by ldez** [comment]
Using AST means no line: because AST is an abstraction of the code not of the line.

So creating a linter on this topic based on AST is not trivial because it requires handling several unrelated things at the same time.

**Comment by mitar** [comment]
> because AST is an abstraction of the code not of the line.

Sure, but every token in AST tree has line number, line offset, and line end. So you just go over all tokens and check for all tokens where they end. If they end after the max line length position, you issue lll warning for that line? And you can then skip struct tag nodes while walking the AST?

**Comment by ldez** [comment]
just an example of the complexity:
```
const aaa = `foofoofoofoofoofoo` + "`foofoofoofoofoo`" + `foofoofoofoofoofoo`
```

**Comment by fedorkanin** [comment]
As far as I can see, https://github.com/golangci/golangci-lint/pull/3986 wasn't merged as it's very specific. As URLs can be different (may not even be HTTP), and comments featuring them do not always start with the URL itself, I suggest a rule for the ones seeking a generalized solution:

```
  exclusions:
      - linters: [lll]
        source: '^\s*//.*\S{80,}'
```

This rule ignores single-line comments that contain an 80+ characters word (often URL). The number can be replaced to one's liking.
